### ISSUE TITLE
Writing tests first and implementing later

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/testing#4]
It is a constant problem that I write my tests and later I need to skip all of them except one I am working on. I need a way to skip tests that are not implemented yet. Typing skip to the test files one by one what most of the testing frameworks do, but I think that is the worst way to do it and it is error prone.

**Comment by inf3rno** [comment]
What I really need is writing the tests first and after that writing the runner file which cares about which tests to run, etc. I could call it coordinator too. Currently the loader decides which tests to load. Maybe it could be the loaders job to solve this skipping problem. I thought of something that follows the natural thought process. Currently we have include exclude in the loader as far as I remember and we don't do skipping. There is auto skip only if a dependency does not work. So I should solve this with exclude. Maybe even remove the skip word from the project and do something like `{failed, reason: DependencyFailed}`. First I want to finish what I originally wrote and refactor after that. I won't realease it until this is solved, so this is not a big deal, I can wait.

**Comment by inf3rno** [comment]
What we have current is this: 
```js
var selection = new Selection();
selection.include(new FileNameFilter("./test/**/spec.*.js"));
var loader = new ESMLoader();
var plan = loader.createExecutionPlan(selection);
```

**Comment by inf3rno** [comment]
We use selections to store the filters for selecting actual tests. After the selection is made we use the loaders to load all tests matching the selection. We can include and exclude tests using the selection. The selection object is mutable.

```
var selection = new Selection();
selection.include(new FileNameFilter("./test/**/*.js"));
selection.exclude(new FileNameFilter("./test/utility/*"));
selection.exclude(new FileNameFilter("./test/helper/*"));
```

**Comment by inf3rno** [comment]
Currently a selection knows only AND or NOT. Adding support for OR would make it very complex and I don't want to invent a query language, but we can do an upper level OR which is sufficient by loading multiple selections.

**Comment by inf3rno** [comment]
As of the skipping I think we could include the test name we are working on and exclude everything else. While that could work it would be nice to do some sort of implemention detection. We could throw a special not implemented error in our implementation, but even that requires some work. Another option to detect somehow the not defined errors and skip those tests automatically, but that is error prone, because we cannot be sure that it is not implemented yet or something is undefined because of a bug.

**Comment by inf3rno** [comment]
Ok, how do we do this in practice?

 - We start to work on a feature.
 - We do something like `include("./test.features.js")`.
 - Next we find our related test `include("./test.features.js#my feature")`.
 - We implement it, it goes green, we step forward for the next test.
 - We need to include that somehow too.
 - Next we find our related test `include("./test.features.js#my feature#another feature")`.
 - And so on...

**Comment by inf3rno** [comment]
This requires some extra work, so better to do it with series or ranges, something like:

```js
var selection = new Selection();
selection.include(new FileNameFilter("./test.features.js"));
selection.include(new TestRangeFilter(0, "another feature"));
```

This should load all tests from the first test to the "another feature" test.

**Comment by inf3rno** [comment]
While this is fine it is too ad-hoc for my taste and my main problem that it is not clear if we select this range in all files we included or just in a single file. I think we should be able to select a single file. So maybe we should do `select.file("./test.features.js").testRange(select.firstTest, "another feature");` This sounds somewhat better. We can do `.but()` if we want to exclude something. So there should be a hierarchy. First we select the files, next we select the tests. I think we should make these immutable too, so we can reuse a file selection and be able to do something like:
```js
var featuresFile = select.file("./test.features.js");
var testRange = featuresFile
    .testRange(featuresFile.firstTest(), "another feature")
    .except(featuresFile.test("complicated feature"));
var loader = new ESMLoader();
var plan = loader.createExecutionPlan(testRange);
```
Now this looks like my style. It looks great.

**Comment by inf3rno** [comment]
I still want to support custom filters, which won't be easy if we make this hierarchical. Not sure if that is the right word for it, but I don't have better. What I mean is first selecting the files and selecting the tests or test groups only after that.

**Comment by inf3rno** [comment]
Ok let's think about it whether we can do the same with
```js
selection
    .include(fileFilter, testFilter)
    .exclude(fileFilter, testFilter)
```
and
```js
selection
    .include(fileFilter)
    .exclude(fileFilter)
    .include(testFilter)
    .exclude(testFilter)
```

**Comment by inf3rno** [comment]
I think they result the same and they can be implemented the same way if we don't want to load the tests just creating a query.

**Comment by inf3rno** [comment]
So what I really did now is not something hierarchical, but it is just making the code a lot more reusable. Making selections immutable has the same effect.

**Comment by inf3rno** [comment]
We could have a plugin system for the selection if we want to add new methods, but I like the filter approach, because it is OOP. I'd rather do `selection.include(Filter.file(...))` than doing `selection.includeFile(...)`. I know there is nothing wrong with `selection.include(new FileNameFilter(...))` either, but now that everybody uses factories and containers it just does not feel right.

**Comment by inf3rno** [comment]
For the sake of simplicity I think I'll stick with the `selection.include(new CustomFilter(...))` approach. One drawback that we need to import every filter we want to use one by one or we do `import *`, but `require` does not support that. Maybe we should have `Filter.file(...)` after all because that can load all filters, but we have a `FileNameFilter` and a `TestFilter` currently and I don't want a plugin system for filters to be able to do `Filter.my(...)`, so I rather skip that part. We can still do `FileNameFilter.create(...)` if we want to avoid the `new` keyword, but I want to keep it simple, so I go with the `new`.

**Comment by inf3rno** [comment]
Another thing I realized meanwhile that we should be able to pass both selections and filters to the include and exclude methods. I did it that way already without realizing it. Now I know what I was doing by instinct, so I can make it compatible with our API.

```js
var featuresFile = select.file("./test.features.js");
var testRange = featuresFile
    .testRange(featuresFile.firstTest(), "another feature")
    .except(featuresFile.test("complicated feature"));
```

```js
var featuresFile = select(new FileNameFilter("./test.features.js"));
var testRange = featuresFile
    .select(new TestRangeFilter(
        featuresFile.select(new FirstTestFilter()), 
        new TestNameFilter("another feature")
    ))
    .exclude(featuresFile.select(new TestNameFilter("complicated feature")));
```

**Comment by inf3rno** [comment]
I think this is easy to understand. I had to remove the `include` word, because one could misunderstand it and think that the selection is mutable. We narrow down the selection with `select`. Even with that word it is not obvious. The `exclude` can stay or we can use `skip` instead of it. As of the `select` I thought about using `reduce` or `narrow`. Maybe `reduceDownTo` would be the right expression but it is too long. I need to check existing libraries or applications which work with database selections and do this narrowing approach. I know such a system and I want to implement it in my knowledge base project too, so it will be useful to learn about the existing solutions.

**Comment by inf3rno** [comment]
I think the `featuresFile` is a wrong variable name here, what we really have is a selection of all tests from that file. So it should be `allTestsFromFeaturesFile` or `allTests` or `allFeatureTests` or `featureTests`. With that variable the `reduce` and the `narrow` works too.

**Comment by inf3rno** [comment]
```js
var featureTests = select(new FileNameFilter("./test.features.js"));
var testRange = featureTests
    .reduce(new TestRangeFilter(
        featureTests.reduce(new FirstTestFilter()), 
        new TestNameFilter("another feature")
    ))
    .skip(featureTests.reduce(new TestNameFilter("complicated feature")));
```

**Comment by inf3rno** [comment]
We can use an alternative selection wrapper for plugins:

```js
var featureTests = x.selectFile("./test.features.js");
var testRange = featureTests
    .selectTestRange(
        featureTests.selectFirstTest(), 
        "another feature"
    )
    .skip(featureTests.selectTests("complicated feature"));
```

**Comment by inf3rno** [comment]
One thing here is not evident. I use context by filtering.

```js
featureTests
    .selectTestRange(
        featureTests.selectFirstTest(), 
        "another feature"
    )
```

vs 

```js
featureTests
    .selectTestRange(
        featureTests.selectFirstTest(), 
        featureTests.selectTests("another feature")
    )
```

**Comment by inf3rno** [comment]
So if I pass the filter, it always uses the context of the called method. In the case of `selectTestRange` it is obvious that we speak about a selection by name if we give a string and not about a selection by file name or something else. We can use the same approach by skip too.

**Comment by inf3rno** [comment]
```js
var featureTests = x.selectFile("./test.features.js");
var testRange = featureTests
    .selectRange(
        featureTests.first(), 
        "another feature"
    )
    .skip("complicated feature");
```

In this case we can do this, because we selected only a single file. If we select multiple files, we need to write `skipTests` and `selectTestRange`.

**Comment by inf3rno** [comment]
With a constant we can even remove the `featureTests.first()` because it is in the context.

```js
var selection = Query
    .selectFile("./test.features.js")
    .selectRange(Query.first, "another feature")
    .skip("complicated feature");
```

**Comment by inf3rno** [comment]
I think we can keep it this way if we do something like `selectFileRange` and `skipFile` and simple `select` for selecting tests by name.

**Comment by inf3rno** [comment]
One thing I noticed meanwhile that we wanted to use `include` for two different things. One is reduction and the other is addition. Or I could say AND and OR. The latter one is not supported currently and it was hard to find the proper word, because there is no word which can express both things at once. So we need another method:

```js
var selection = Query
    .selectFile("./test.features.js")
    .narrowRange(Query.first, "another feature")
    .skip("complicated feature")
    .merge(Query.selectFile("./test.moreFeatures.js"));
```
In practive we can solve this with 3 words AND, NOT, OR or maybe `reduce`, `include`, `exclude`.

**Comment by inf3rno** [comment]
```js
var featuresFile = select(new FileNameFilter("./test.features.js"));
var testRange = featuresFile
    .select(new TestRangeFilter(
        featuresFile.select(new FirstTestFilter()), 
        new TestNameFilter("another feature")
    ))
    .exclude(featuresFile.select(new TestNameFilter("complicated feature")));
```

vs

```js
var tests = selection
    .include(new FileNameFilter("./test.features.js"))
    .reduce(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"));
```

**Comment by inf3rno** [comment]
Meanwhile I realized that `.exclude(featuresFile.select(new TestNameFilter("complicated feature")));` is using the wrong selection because we already reduced that selection to a more narrow one.

**Comment by inf3rno** [comment]
I think the upper that is using only filters is completely acceptable. Of course we should be able to use selections too, but we should avoid them for the sake of simplicity if we use the same context for the selection as we use for the method call.

**Comment by inf3rno** [comment]
I am not sure that the include, reduce, exclude words are the best here, but for now they are ok. I am thinking on loading the actual modules. In that case I should run the upper code with the loader so I could write.

```js
var tests = loader
    .select(new FileNameFilter("./test.features.js"))
    .reduce(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"));
```

What I do by reduce is subsetting. I need to check if the typical reduce contains filter for the selection or the exclusion. If the latter I'll need to find a different word. Maybe include, select, exclude.

**Comment by inf3rno** [comment]
The thing is reduce is used more or less to make aggregates of collections or to change the collection structure. They can be used for filtering too, but they typically need to build some sort of "total", which can be the filtered array too ofc. Better to avoid that word imho.

**Comment by inf3rno** [comment]
The filter method is used for subsetting in js, so that would be a nice name. I just never liked it because it is not obvious if we want to exclude or narrow down the selection with the filter. Maybe we could use match instead, but that would be weird.

**Comment by inf3rno** [comment]
```js
var tests = loader
    .include(new FileNameFilter("./test.features.js"))
    .filter(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"));
```

**Comment by inf3rno** [comment]
Can we use the `include` with a `TestNameFilter`?
I think we can.

```js
var tests = loader
    .include(new FileNameFilter("./test.features.js"))
    .filter(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"))
    .include(new TestNameFilter("complicated feature"))
```
But it is not obvious what is the context here. We can use the filter results or the include results as context. If we use the exclude result as context, then we simply cannot include anything more except if we make an implicit rule, that in these cases all the included files are the context.

I don't want to have implicit rules, so the latter should work:

```js
var allTests = loader.include(new FileNameFilter("./test.features.js"));
var selectedTests = allTests
    .filter(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"))
    .include(allTests.filter(new TestNameFilter("complicated feature")));
```

**Comment by inf3rno** [comment]
Of course we can avoid the upper by passing multiple filters, but this looks nice too.

**Comment by inf3rno** [comment]
I wanted to support `TestFilter` too, which can filter by imported tests directly.
```js
import {myTest} from "./test.features.js";
new TestFilter(myTest);
```
I am not sure if this is necessary or beneficial. I mean we already have our filters for test name we can simple add a filter for variable name and problem solved. Another issues with it, that it makes more complicated to replace the loader in these runner files, because we need to rewrite these too. And yet another reason that if we do it thise way, then we don't even need loaders and selectors, just importing them one by one and problem solved. That was my original plan actualy, but the import system is not good enough to do it that way if I have a long list of tests...

**Comment by inf3rno** [comment]
Actually lazy loading and using filters only is not necessary. Lazy loading is useful in this case if we want to skip entire files, but the file content must be loaded if we filter individual tests. It is still good if we want to optimize it e.g. in the upper example merging the two filters if possible. Actually I was wrong and it is not possible to change the include - exclude order in that example.

**Comment by inf3rno** [comment]
So I use the loader because the import system is not good enough to import files by regex and I want to skip the tests with selectors instead of writing the skips into the test files. Another reason I wanted to use the loader to make group and test instances from object literals.

**Comment by inf3rno** [comment]
I think the loader should still be compatible with manual import and manual test selection.

**Comment by inf3rno** [comment]
We still have one thing here. When we first use include we pass a file name filter so the loader goes through file names. This means that we'll need at least two custom filter types, one for file names and one for tests. Otherwise the loader won't know what to do with it by filter type and would try to run the test filter through file names. I think that solves our issue with files first tests second...

**Comment by inf3rno** [comment]
Another option is using the same filter but giving the selection a list of file names instead of a list of tests. That is possible, we can even have `FileName` objects for that and do something like `if (item instanceof FileName) ...` in our tests. On the other hand this means that each time we do file name filtering we run our tests through the test names and each time we do test filtering we run the tests through the file names. This makes the selection slow. On the other hand we can reduce the number of filters with it.

```js
var tests = loader
    .include(new NameFilter(FileName, "./test.features.js"))
    .filter(new RangeFilter(Test, RangeFilter.first, new NameFilter(Test, "another feature")))
    .exclude(new NameFilter(Test, "complicated feature"))
```

**Comment by inf3rno** [comment]
We have a problem with this approach. It is ugly config instead of code, I'd rather use instanceof. And in the case of our range we need to pass it to every subfilter. I could make the RangeFilter smart enough to pass them though.

**Comment by inf3rno** [comment]
What if we start to write code?

```js
var tests = loader
    .filter(function (selection, file){
        if (file.path === "./test.features.js")
            selection.add(file);
    })
    .load()
    .filter(function (selection, test, iterator){
        if (test.name !== "complicated feature")
            selection.add(test);
        if (test.name === "another feature")
            iterator.stop();
    });
```

**Comment by inf3rno** [comment]
I love to write code in my config files too and hate to create config objects. I understand that the latter is very useful, because unlike the code I can serialize it, but in many cases writing a few lines of code is so much simpler...

**Comment by inf3rno** [comment]
```js
var loader = new Loader();

var selectedFiles = [];
for (var file of loader.loadDirectory("./tests"))
    if (file.path === "./test.features.js")
        selectedFiles.push(file);
		
var selectedTests = [];
for (var test of loader.loadFiles(selectedFiles)){
	if (test.name !== "complicated feature")
		selectedTests.push(test);
	if (test.name === "another feature")
		break;
}
```

Of course with the upper approaches the iteration and the comparison is not reused which makes it a lot more low level than 

```js
var tests = loader
    .include(new FileNameFilter("./test.features.js"))
    .filter(new TestRangeFilter(TestRangeFilter.first, new TestNameFilter("another feature")))
    .exclude(new TestNameFilter("complicated feature"));
```

and we cannot do lazy load with query optimization etc...

**Comment by inf3rno** [comment]
Still it is very simple and easy to understand at least for a programmer... We have to import only a single Loader class and write two loops....

**Comment by inf3rno** [comment]
We can simplify the upper:
```js
var tests = loader
    .filter("./test.features.js")
    .load()
    .filter([0, "another feature"])
    .exclude("complicated feature");
```

**Comment by inf3rno** [comment]
If we separate our file and test filters.

```js
var files = selection.filter("./test.features.js")
var tests = selection.filter([0, "another feature"]).exclude("complicated feature");
loader.load(files, tests);
```

I am not sure if this is the right way.

**Comment by inf3rno** [comment]
I think I want to write flexible code because I don't know the use cases. I don't know how others use testing libraries. I need to think about it. For example tagging tests or assigning them to people would be more useful than finding tests by name.

**Comment by inf3rno** [comment]
This depends on #9 . I would rather load most of the modules and check if they contain anythat that related to the Test class. But this is just an idea. It might depend on #15 if we tag our tests. So this is a complicated topic. I'd rather write code instead of using a CLI parameter, which equals to configuration.

**Comment by inf3rno** [comment]
I think this has the same solution as #15 or at least there are things that overlap. We can follow a load all and filter later approach, though it would be somewhat slow and it would not support testing the environment with syntactically invalid code. For example using es6 syntax in an es5 environment. Though we can add eval or new Function to avoid issues. I think this will be hard.

Probably we need to use a multi level approach. First filter is the path of the js file, second filter is the class name, which can be the same as the file name, though a single file can contain multiple classes. Third is the function name. Fourth is the tag if we give one. We can use the same by skipping tests. We should be able to run the skip filter along with the selection filter with an include exclude approach. Though it is a little bit complicated because of #11. I guess we should use a filter object rather than a filter function, for example we have this in our previous code:

```js
var tests = loader
    .include(new NameFilter(FileName, "./test.features.js"))
    .filter(new RangeFilter(Test, RangeFilter.first, new NameFilter(Test, "another feature")))
    .exclude(new NameFilter(Test, "complicated feature"))
```

As far as I understand (I don't want to read back too much), we have a range from the first test to the test called "another feature" in our test file we want to run and we have a test called "complicated feature" we want to skip. Using test levels would look like this:

```js
var tests = loader({
    path: function (path){
        return path == "./test.features.js";
    },
    "class": function (test){
        return (test instanceof Test);
    },
    include: true,
    "function": function (func, index){
        if (func.name == "another feature") 
            this.include = false;
        return this.include;
    },
    tag: function (tags){
        if (tags.has("complicated"))
            return false;
        return true;
    }
});
```

**Comment by inf3rno** [comment]
Normally we do the select, exclude, filter approach, but that is good for configuration. It means that first we select every test we want to run, next we run them in another loop by passing it to a different class. While this is ok, I don't like the approach. If we use the same code for skipping, then we need to pass the skip selectors in the beginning and run them along with the normal selectors. So we should be able to load normal selectors first and load the skip selectors second.

It would be a very nice solution to do the actual selection right before the testing. So whenever we find a test we execute it and move to the next test. This way we can use the loader to check things like class name and function name too instead of splitting up selection logic into multiple classes or making it complicated. Actually we don't even need to use the skip word, we just need to add another exclusion rule. So this would simplify things, because we can use the exact same code only once for both selection and skipping. I think this is amazing. As of the tagging I would keep them, they can be useful, but we need a way to tag individual functions, which is hard. Tagging objects is easy, we just add a set of tags to the object.

**Comment by inf3rno** [comment]
I think the original 

```js
selection
    .include(fileFilter, testFilter)
    .exclude(fileFilter, testFilter)
```

was close to our real solution.

What we really need is something like

```js
var findAll = selection.include({fileFilter(){}, testFilter(){}});
var skipSome = selection.exclude({fileFilter(){}, testFilter(){}});
runner.select(findAll, skipSome).execute(p,q,r);
```

Actually we can do the selection plan approach too with lazy loading:

```js
var findAll = selection.include({fileFilter(){}, testFilter(){}});
var skipSome = selection.exclude({fileFilter(){}, testFilter(){}});
var plan = loader.select(findAll, skipSome)
var report = executor.execute(plan, [p,q,r]);
reporter.display(report);
```

Here we can use an event storage as report, so we can use multiple reporters to display it any time without having a real issue about when we start it.

**Comment by inf3rno** [comment]
I closed the #15, because it is solved mostly. I'll open a new issue for the tagging part.

**Comment by inf3rno** [comment]
One thought here, that it is not obvious how we merge multiple selectors. We can use an AND or an OR operator too. In the upper example we used AND, but in many cases we'll need OR.

**Comment by inf3rno** [comment]
```js
var plan = loader.select(findAll, skipSome);
```
Here we use AND implicitly. What can we do about this?

```js
var plan = loader.select(selector.and(findAll, skipSome));
```
```js
var plan = loader.select(operator.and(findAll, skipSome));
```
```js
var plan = loader.select(findAll.and(skipSome));
```
```js
var plan = loader.select(findAll).and(skipSome);
```

**Comment by inf3rno** [comment]
I think the first 3 is ok, the last one is bad, because the loader is not responsible for building the selector. It should just get a selector that can decide if we want to load a certain file and run a certain test from that file.

**Comment by inf3rno** [comment]
Currently we have an `isTest` function in our code in. #8 

```js
    getTestNames(){
        var reservedNames = Object.getOwnPropertyNames(Test.prototype);
        var testNames = new Set();
        walkPrototypeChain(this, function (object){
            if (object === Object.prototype || object === Test.prototype)
                return;
            for (var property of Object.getOwnPropertyNames(object))
                if (reservedNames.indexOf(property) == -1 && (this[property] instanceof Function) && util.types.isAsyncFunction(this[property]))
                    testNames.add(property);
        }.bind(this));
        return testNames;
    }
```

I am thinking on moving that code to our loader or selector. Currently it is an implicit rule that all async functions without reserved names are tests. We can make that explicit by moving it to the loader, though in that case having a Test class would be meaningless I think and we would end up with data structures instead of objects. So we should decide if the Test should be responsible for the execution of its content or it should not exist at all. Good question. It causes many problems actually, because our function tagging #17 is a consequence of having it more or less. In our previous version we had individual tests in our test class each containing a single function called execute, so it was not hard to tag it. On the other hand our current approach is a lot better for how we normally test. Actually in many cases we use even deeper nested structures with describe + it. I'll need to check some examples to decide if we want to support such things.
